src/wp-includes/media.php

```diff
@@ -1958,6 +1981,103 @@ function wp_img_tag_add_loading_attr( $image, $context ) {
+ */
+function wp_img_tag_add_loading_optimization_attrs( $image, $context ) {
+	$image_elm = new WP_HTML_Tag_Processor( $image );
+	if ( ! $image_elm->next_tag() ) {
+		return $image;
+	}
+
+	$width             = $image_elm->get_attribute( 'width' );
+	$height            = $image_elm->get_attribute( 'height' );
+	$loading_val       = $image_elm->get_attribute( 'loading' );
+	$fetchpriority_val = $image_elm->get_attribute( 'fetchpriority' );
+
+	// Bail early if both the priority attributes are already present.
+	if ( ! empty( $loading_val ) && ! empty( $fetchpriority_val ) ) {
+		return $image;
+	}
+
+	$optimization_attrs = wp_get_loading_optimization_attributes(
+		'img',
+		array(
+			'width'         => $width,
+			'height'        => $height,
+			'loading'       => $loading_val,
+			'fetchpriority' => $fetchpriority_val,
+		),
+		$context
+	);
+
+	// Retained for backward compatibility.
+	$loading_attrs_enabled = wp_lazy_loading_enabled( 'img', $context );
+
+	if ( empty( $loading_val ) && $loading_attrs_enabled ) {
+		/**
+		 * Filters the `loading` attribute value to add to an image. Default `lazy`.
+		 * This filter is added in for backward compatibility.
+		 *
+		 * Returning `false` or an empty string will not add the attribute.
+		 * Returning `true` will add the default value.
+		 * `true` and `false` usage supported for backward compatibility.
+		 *
+		 * @param string|bool $optimization_attrs['loading'] Current value for `loading` attribute for the image.
+		 * @param string      $image   The HTML `img` tag to be filtered.
+		 * @param string      $context Additional context about how the function was called or where the img tag is.
+		 */
+		$filtered_loading_attr = apply_filters(
+			'wp_img_tag_add_loading_attr',
+			isset( $optimization_attrs['loading'] ) ? $optimization_attrs['loading'] : false,
+			$image,
+			$context
+		);
+
+		// Validate the values after filtering.
+		if ( isset( $optimization_attrs['loading'] ) && ! $filtered_loading_attr ) {
+			// Unset `loading` attributes if `$filtered_loading_attr` is set to `false`.
+			unset( $optimization_attrs['loading'] );
+		} elseif (
+			( isset( $optimization_attrs['loading'] ) ? $optimization_attrs['loading'] : false ) !== $filtered_loading_attr &&
+			// fetchpriority='high' and loading='lazy' should be mutually exclusive.
+			'lazy' === $filtered_loading_attr && 'high' === $optimization_attrs['fetchpriority']
+		) {
+			_doing_it_wrong(
+				__FUNCTION__,
+				sprintf(
+					/* translators: %s: fetchpriority="high". */
+					__( 'An image cannot be lazy-loaded and assigned %s at the same time.' ),
+					'<code>fetchpriority="high"</code>'
+				),
+				'6.3.0'
+			);
+
+			// The filtered value will still be respected.
+			$optimization_attrs['loading'] = $filtered_loading_attr;
+		} elseif ( in_array( $filtered_loading_attr, array( 'lazy', 'eager' ), true ) ) {
+			$optimization_attrs['loading'] = $filtered_loading_attr;
+		}
```

Comment:
It would be simpler to have the new specific condition around the `_doing_it_wrong()` as a subcondition of the main `if ... elseif ...`, which is what I meant in https://github.com/WordPress/wordpress-develop/pull/4495#discussion_r1213449575. The way you have it now, it's more complicated to follow, and lines 2064 and 2066 are duplicated.
```suggestion
		} elseif ( in_array( $filtered_loading_attr, array( 'lazy', 'eager' ), true ) ) {
			/*
			 * If the filter changed the loading attribute to "lazy" when a fetchpriority attribute
			 * with value "high" is already present, trigger a warning since those two attribute
			 * values should be mutually exclusive.
			 */
			if (
				( isset( $optimization_attrs['loading'] ) ? $optimization_attrs['loading'] : false ) !== $filtered_loading_attr &&
				'lazy' === $filtered_loading_attr && 'high' === $optimization_attrs['fetchpriority']
			) {
				_doing_it_wrong(
					__FUNCTION__,
					__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
					'6.3.0'
				);
			}

			$optimization_attrs['loading'] = $filtered_loading_attr;
		}
```

Reply:
Made the change.